src/sync.ts

```diff
@@ -239,6 +247,37 @@ export class SyncApi {
                 RoomMemberEvent.Membership,
             ]);
         });
+
+        room.currentState.on(RoomStateEvent.Marker, function(event, {fromInitialState}: ISetStateOptions = {}) {
+            // We don't want to refresh the timeline:
+            //  1. If it's persons first time syncing the room, they won't have
+            //     any old events cached to refresh.
+            //  1. If we're re-hydrating from `syncFromCache` because we already
+            //     processed any marker event state that was in the cache
+            if(fromInitialState) {
+                console.log('fromInitialState ignoring');
+                return;
+            }
+
+            const isValidMsc2716Event =  MSC2716_ROOM_VERSIONS.includes(room.getVersion()) ||
+                // MSC2716 is supported in all existing room versions but special
+                // meaning should only be given to "insertion", "batch", and
+                // "marker" events when they come from the room creator
+                event.getSender() === room.getRoomCreator();
+            console.log(`On marker state isValidMsc2716Event=${isValidMsc2716Event}, event_id=${event.getId()} room.getLastMarkerEventIdProcessed()=${room.getLastMarkerEventIdProcessed()}`);
+            if(
+                isValidMsc2716Event &&
+                // We only need to throw the timeline away once, when we see a
+                // marker. All of the historical content will be in the
+                // `/messsages` responses from here on out.
+                event.getId() !== room.getLastMarkerEventIdProcessed()
+            ) {
+                // Saw new marker event, refreshing the timeline
+                console.log('Saw new marker event, refreshing the timeline', new Error().stack);
+                room.resetLiveTimeline(null, null);
```

Comment:
TODO: Test if this is the function we actually want to use.

Our goal is to throw away the current timeline and re-fetch events for the room from the server.

Reply:
It would probably be better if we re-fetched the new events first, then replaced the timeline

Reply:
Thinking about this more, do we even want to do this automatically? Someone could cause your Element client to constantly refresh the timeline by just sending marker events over and over. Granted, you probably want to leave a room like this 🤷. Perhaps also some sort of DOS vector since everyone will be refreshing and hitting the server at the exact same time.

I think we might want to approach this to be a similar flow to how Element receives app updates. When the client receives a marker event, we first notify the user that history was imported and is available to see after they refresh their timeline (which can be a button action). The UI can probably use something similar to the "Connectivity to the server has been lost." banner.

<img width="701" alt="" src="https://user-images.githubusercontent.com/558581/162538136-0a947a91-3fdd-48ce-833c-64084f2917ff.png">

Reply:
Updated to prompt the user to refresh their timeline: https://github.com/matrix-org/matrix-react-sdk/pull/8303

<img width="928" alt="" src="https://user-images.githubusercontent.com/558581/163079107-fdc250d4-5b3e-4d4b-9ff8-70714cc65c72.png">



> We can detect whether the history imported affects any timelines on the client and only ask to refresh then. To detect, when we receive a marker event, make a request to `/context/{insertionEventId}` with the `insertionEventId` the marker is pointing to and check if any its `prev_events` are in the timeline. `/context` doesn't work over federation but the homeserver backfills the insertion event when it receives the marker so it already has it available.

We can't do this because `prev_events` aren't available in the client API's 😢. I'm not too keen to add an extra field to the marker or insertion event indicating where the history is imported next to as there isn't a way to enforce that it's true (a homeserver could lie and make it up). I'd rather keep the source of truth in the `prev_events`.